### ISSUE TITLE
[BUG] Consolidation is perfomed before function is evaluated resulting in wrong aggregated results

### ISSUE THREAD

**Issue by limpasha** [go-graphite/carbonapi#611]
**Describe the bug**
When querying for data for a long period with any aggregating function (like summarize, for example), incorrect result are returned, which is unexpected and so this types of queries seems unreliable. As far as I understand the root cause of the problem is adjusting `step` according to `maxDataPoints`.

**Detailed explanation**
Imagine the query `target=summarize(pulse.test_sum, "1d", "sum", false)&from=-10d&until=now&maxDataPoints=100`. The right way to do it correctly is to
- fetch **all** data points from storage
- compute 10 sums, which results in 10 points (10 days, summarize by 1 day)

**So**, what carbonapi does here https://github.com/go-graphite/carbonapi/blob/bb33f26c8466deb9c9134c1085d6fcf77cc02205/cmd/carbonapi/http/render_handler.go#L268
- fetches datapoints https://github.com/go-graphite/carbonapi/blob/bb33f26c8466deb9c9134c1085d6fcf77cc02205/expr/expr.go#L69
- inside the function above `step` **is adjusted according to** `maxDataPoints` provided param, that's why not all metrics are returned from backend https://github.com/go-graphite/carbonapi/blob/bb33f26c8466deb9c9134c1085d6fcf77cc02205/zipper/protocols/prometheus/prometheus_group.go#L236
- then this **incomplete data is used in aggregation** https://github.com/go-graphite/carbonapi/blob/bb33f26c8466deb9c9134c1085d6fcf77cc02205/expr/expr.go#L100

**CarbonAPI Version**
bb33f26c8466deb9c9134c1085d6fcf77cc02205

**Logs**
1. Data we have prepared (show only non-null), maxDataPoints= 10days * 60s step:
```
 % curl '0:8081/render?target=pulse.test1_sum&format=csv&from=-10d&until=now&maxDataPoints=14400' -s | egrep -v ',$'"pulse.test1_sum",2021-05-08 14:19:00,1
"pulse.test1_sum",2021-05-08 14:20:00,1
"pulse.test1_sum",2021-05-08 14:21:00,1
"pulse.test1_sum",2021-05-09 14:19:00,1
"pulse.test1_sum",2021-05-09 14:20:00,1
"pulse.test1_sum",2021-05-09 14:21:00,1
"pulse.test1_sum",2021-05-09 14:22:00,1
"pulse.test1_sum",2021-05-09 14:23:00,1
"pulse.test1_sum",2021-05-10 14:19:00,1
"pulse.test1_sum",2021-05-11 14:20:00,1
"pulse.test1_sum",2021-05-11 14:21:00,1
"pulse.test1_sum",2021-05-12 14:20:00,1
"pulse.test1_sum",2021-05-12 14:21:00,1
"pulse.test1_sum",2021-05-12 14:22:00,1
"pulse.test1_sum",2021-05-12 14:23:00,1
"pulse.test1_sum",2021-05-12 14:24:00,1
"pulse.test1_sum",2021-05-12 14:25:00,1
"pulse.test1_sum",2021-05-12 14:26:00,1
"pulse.test1_sum",2021-05-12 14:27:00,1
"pulse.test1_sum",2021-05-12 14:28:00,1
```
2. Then try `summarize` with **enough** datapoints (_correct_):
```
% curl '0:8081/render?target=summarize(pulse.test1_sum,"1d","sum",false)&format=csv&from=-10d&until=now&maxDataPoints=14400' -s | egrep -v ',$'
"summarize(pulse.test1_sum,'1d','sum',false)",2021-05-08 00:00:00,3
"summarize(pulse.test1_sum,'1d','sum',false)",2021-05-09 00:00:00,5
"summarize(pulse.test1_sum,'1d','sum',false)",2021-05-10 00:00:00,1
"summarize(pulse.test1_sum,'1d','sum',false)",2021-05-11 00:00:00,2
"summarize(pulse.test1_sum,'1d','sum',false)",2021-05-12 00:00:00,9
```
3. Then try `summarize` with **NOT enough** datapoints (_incorrect_):
```
% curl '0:8081/render?target=summarize(pulse.test1_sum,"1d","sum",false)&format=csv&from=-10d&until=now&maxDataPoints=1440' -s | egrep -v ',$' 
"summarize(pulse.test1_sum,'1d','sum',false)",2021-05-08 00:00:00,1
"summarize(pulse.test1_sum,'1d','sum',false)",2021-05-09 00:00:00,1
"summarize(pulse.test1_sum,'1d','sum',false)",2021-05-11 00:00:00,1
"summarize(pulse.test1_sum,'1d','sum',false)",2021-05-12 00:00:00,1
```

**PS**

As far as I have dived into the problem deeply, I thinks I would be able to make PR fixing it. My idea is to have at least some configuration flag which forces using minimal `step` when fetching data. What do you guys think about it?

**CarbonAPI Configuration:**
```
listen: "0.0.0.0:8081"
concurency: 20
cpus: 10
tz: ""
sendGlobsAsIs: true
maxBatchSize: 1000000
cache:
    type: "null"
idleConnections: 10
pidFile: ""
upstreams:
    buckets: 10
    timeouts:
        global: "60s"
        afterStarted: "10s"
        connect: "200ms"
    concurrencyLimit: 10
    keepAliveInterval: "30s"
    maxIdleConnsPerHost: 100
    tldCacheDisabled: true
    backends:
    backendsv2:
        backends:
          - groupName: "victoria"
            protocol: "victoriametrics"
            lbMethod: "broadcast"
            backendOptions:
                step: "60"
            servers:
                - "http://dev:5000"
expireDelaySec: 10
```

**Comment by Civil** [comment]
That is not exactly a bug per-se. There are several reasons for that behavior:

1. When you don't have enough width to display a graph, there is no point in fetching precise data. It is actually a responsibility of whoever send data to set correct aggregation function or whoever define the dashboard to set correct `consolidateBy`.
2. Browsers are not really great at doing a lot of math on top of a huge pile of data, especially when you have a bunch of graphs in Grafana on a single dashboard. That was the whole point behind maxDataPoints in a first place (and here I mean in graphite-web).
3. Operating on a lot of points is not free (basically doesn't scale well, as you'll need to scale up your frontends with carbonapi as well as amount of backends, and in case of fast storage engines, like victoriametrics or clickhouse it is much better to do some operations on backend).

And while aggregations on aggregated data is not entirely correct, it is still possible to get idea of the problems of the data behind it (however ability to investigate depends on consolidation function on the backend and frontend). However I'm OK if you'll introduce a config flag that will allow to turn off this behavior.

There are actually two other possibilities how to solve this (but much harder than a config option to disable that behavior completely):

1. Create a list of functions that can be applied on a backend level and pass them to the backend.
2. Allow function to specify if it's ok to accept aggregated data from a backend and if it by itself changes aggregation status of the data and if any of the functions requires full data set, do not aggregate the data on a backend (or before it reached the `expr`)

**Comment by limpasha** [comment]
Thank you for such a detailed explanation! I totally agree with the reasons why `maxDataPoints` param was initially introduced. And workarounds you've provided are applicable.

As with my case I think configuration flag would be better solution because the precision and correctness of aggregated results is required, it is used in my scripts, no always in visualization tools like Grafana. I'm gonna work on it soon.

P.S. There is even workaround that can be implemented without any changes: after #607 there is possibility to calculate `maxDataPoints` according to provided `from` and known `step` and pass it as param to a request! But yes, it won't help with Grafana case, because Grafana calculates maxDataPoints by itself.
